### PR TITLE
Add support for dictionary with optional/nullable values in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR adds support for dictionaries with optional (nullable) values in C#.

It's similar to the sequence<T?> support added previously. A dictionary<K, V?> is encoded using a bit sequence that indicates which values are null (and therefore not encoded at all). Keys are never optional.

### CODE REVIEW

File: cpp/src/slice2cs/CsUtil.cpp [hunk 1]
```diff
@@ -1089,14 +1089,26 @@ Slice::CsGenerator::writeTaggedMarshalCode(Output& out,
         TypePtr keyType = d->keyType();
         TypePtr valueType = d->valueType();
 
+        bool withBitSequence = false;
+
+        if (auto optional = OptionalPtr::dynamicCast(valueType);
+            optional && !optional->underlying()->isClassType() && !optional->underlying()->isClassType())
```
Comment (bentoi) [inline cpp/src/slice2cs/CsUtil.cpp:1095]: We have this in multiple places, perhaps there's a way to improve this check? Could we extend Optional for this case perhaps?

File: csharp/src/Ice/OutputStream.cs [hunk 2]
```diff
@@ -480,6 +480,200 @@ public void WriteVarULong(ulong v)
...
+                WriteSize(count);
+                BitSequence bitSequence = WriteBitSequence(count);
+                int index = 0;
+                foreach ((TKey key, TValue? value) in v)
+                {
+                    keyWriter(this, key);
+                    if (value is TValue actualValue)
+                    {
+                        valueWriter(this, actualValue);
+                    }
```
Comment (pepone) [inline csharp/src/Ice/OutputStream.cs:509]: I think for reference types we can just use `value` after checking for null
```suggestion
                    if (value != null)
                    {
                        valueWriter(this,value);
                    }
```

File: csharp/src/Ice/InputStream.cs [hunk 3]
```diff
@@ -405,6 +408,70 @@ public uint ReadVarUInt()
...
+        /// <returns>The dictionary read from the stream.</returns>
+        public Dictionary<TKey, TValue?> ReadDictionary<TKey, TValue>(
+            int minKeySize,
+            bool withBitSequence,
+            InputStreamReader<TKey> keyReader,
+            InputStreamReader<TValue> valueReader)
+            where TKey : notnull
+            where TValue : class
+        {
+            int sz = ReadAndCheckSeqSize(minKeySize);
```
Comment (bentoi) [inline csharp/src/Ice/InputStream.cs:426]: Shouldn't this check for the value size if withBitSequence = false? I suppose we don't really use it with false?

File: csharp/src/Ice/InputStream.cs [hunk 4]
```diff
@@ -467,16 +534,19 @@ public object ReadSerializable()
...
-            int minEntrySize,
+            int minKeySize,
+            int minValueSize,
             InputStreamReader<TKey> keyReader,
-            InputStreamReader<TValue> valueReader) where TKey : notnull
+            InputStreamReader<TValue> valueReader)
+            where TKey : notnull
         {
-            int sz = ReadAndCheckSeqSize(minEntrySize);
+            int sz = ReadAndCheckSeqSize(minKeySize + minValueSize);
```
Comment (bentoi) [inline csharp/src/Ice/InputStream.cs:549]: Couldn't we share the same impl for dictionary / sorted dictioannary?

File: cpp/src/slice2cs/CsUtil.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/CsUtil.cpp:1095]: Fixed. It should be ClassType and InterfaceType.

File: cpp/src/slice2cs/CsUtil.cpp [hunk 5]
```diff
@@ -1218,14 +1230,33 @@ Slice::CsGenerator::writeTaggedUnmarshalCode(Output &out,
         assert(d);
         TypePtr keyType = d->keyType();
         TypePtr valueType = d->valueType();
+        bool withBitSequence = false;
+
+        if (auto optional = OptionalPtr::dynamicCast(valueType);
+            optional && !optional->underlying()->isClassType() && !optional->underlying()->isClassType())
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/CsUtil.cpp:1236]: Same.

File: cpp/src/slice2cs/Gen.cpp [hunk 6]
```diff
@@ -2119,6 +2119,15 @@ Slice::Gen::TypesVisitor::visitDictionary(const DictionaryPtr& p)
     string name = p->name();
     TypePtr key = p->keyType();
     TypePtr value = p->valueType();
+
+    bool withBitSequence = false;
+    if (auto optional = OptionalPtr::dynamicCast(value);
+        optional && !optional->underlying()->isClassType() && !optional->underlying()->isClassType())
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:2125]: Fixed.

File: csharp/src/Ice/OutputStream.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/OutputStream.cs:509]: Fixed as suggested.

File: csharp/src/Ice/OutputStream.cs [hunk 7]
```diff
@@ -480,6 +480,200 @@ public void WriteVarULong(ulong v)
...
+            if (withBitSequence)
+            {
+                int count = v.Count();
+                WriteSize(count);
+                BitSequence bitSequence = WriteBitSequence(count);
+                int index = 0;
+                foreach ((TKey key, TValue? value) in v)
+                {
+                    key.IceWrite(this);
+                    if (value is TValue actualValue)
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/OutputStream.cs:576]: Fixed.

File: cpp/src/slice2cs/CsUtil.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/CsUtil.cpp:1095]: Yes, I added a new function `encodedUsingBitSequence`.

File: csharp/src/Ice/InputStream.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:426]: Yes, in theory we should include the minValueSize when withBitSequence is false, but in practice this method is never called with withBitSequence=false. If you want to read a dictionary without a bit sequence, use another `ReadDictionary` overload without this withBitSequence parameter.

File: csharp/src/Ice/InputStream.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:521]: Yes, fixed.

**Pull request merged**